server/src/services/documentSymbol/SymbolVisitor.ts

```diff
@@ -0,0 +1,68 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import { SymbolKind } from "vscode-languageserver-types";
+import _ from "lodash";
+import { RuleKind, TokenKind } from "@nomicfoundation/slang/kinds";
+import { RuleNode, TokenNode } from "@nomicfoundation/slang/cst";
+import { TextDocument } from "vscode-languageserver-textdocument";
+import { Cursor } from "@nomicfoundation/slang/cursor";
+import { slangToVSCodeRange } from "../../parser/slangHelpers";
+import { SymbolTreeBuilder } from "./SymbolTreeBuilder";
+
+export abstract class SymbolVisitor {
+  public abstract ruleKind: RuleKind;
+  public abstract symbolKind: SymbolKind;
+  public abstract nameTokenKind: TokenKind;
+
+  constructor(
+    public document: TextDocument,
+    public symbolBuilder: SymbolTreeBuilder
+  ) {}
+
+  public onRuleNode(node: RuleNode, cursor: Cursor): void {
+    const range = slangToVSCodeRange(this.document, cursor.textRange);
+
+    let symbolName = "-";
+    let selectionRange = slangToVSCodeRange(this.document, cursor.textRange);
+
+    // Find identifier
+    const childCursor = cursor.spawn();
+
+    do {
+      const identifierNode: TokenNode | null = childCursor.findTokenWithKind([
+        this.nameTokenKind,
+      ]);
+
+      if (
+        identifierNode &&
+        _.last(childCursor.pathRuleNodes).kind === this.ruleKind
+      ) {
+        symbolName = identifierNode.text;
+        selectionRange = slangToVSCodeRange(
+          this.document,
+          childCursor.textRange
+        );
+        break;
+      }
+    } while (childCursor.goToNext());
```

Comment:
Question: is there a valid case where the first node we find with `nameTokenKind` is not the intended one?

Reply:
I see. Thanks for explaining!

In that case, if we are checking `cursor.ancestors().length` above, and it turns out it is greater than 1, I suggest calling `cursor.goToParent()` instead of `cursor.goToNext()`, as it will be able to skip the entire subtree.

Reply:
Thanks for reporting! I'm not sure I can reproduce on my end. Can you please send me a test/snippet?
I would be happy to fix it and send you a new build.

Not blocking though.

Reply:
Apologies for the confusion. Looks like we are missing an API on our end. I will review it and get back to you. In the meantime, I think checking `cursor.ancestors().length` should be enough.

Side note: I'm also adding `cursor.parent(): Node` in the next release, to allow you to do both operations in one call. But not blocking at all!

Reply:
Edit: apologies for the confusion.

`goToParent()` would go up, but then continue to revisit that parent, then its children (including the current node) again. What I think we need to do is to `goToParent()` and then `goToNextSibling()`, so when the cursor continues, it skips the entire subtree of the parent (including its children, and the current node).

Is that the behavior you want?

Reply:
Edit: Apologies for the confusion. No need to call `.goToParent()` since you actually want to continue inspecting the remaining children.